### PR TITLE
Migrate ENSIntegration unit tests to Hardhat & Typescript

### PR DESCRIPTION
**Pull request by zguesmi**

### CODE REVIEW

File: test/byContract/ENSIntegration/ENSIntegration.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    ENSRegistry__factory,
+    IexecInterfaceNative,
+    IexecInterfaceNative__factory,
+    PublicResolver__factory,
+    ReverseRegistrar__factory,
+} from '../../../typechain';
+import { getIexecAccounts } from '../../../utils/poco-tools';
+const CONFIG = require('../../../config/config.json');
+
+describe('ENSIntegration', async () => {
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.ts:19]:
```suggestion
describe('ENSIntegration', () => {
```

File: test/byContract/ENSIntegration/ENSIntegration.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    let proxyAddress: string;
+    let iexecPoco: IexecInterfaceNative;
+    let iexecPocoAsAdmin: IexecInterfaceNative;
+    let ensRegistry: ENSRegistry;
+    let iexecAdmin: SignerWithAddress;
+    let anyone: SignerWithAddress;
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.ts:25]: Make it an array ?

File: test/byContract/ENSIntegration/ENSIntegration.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    PublicResolver__factory,
+    ReverseRegistrar__factory,
+} from '../../../typechain';
+import { getIexecAccounts } from '../../../utils/poco-tools';
+const CONFIG = require('../../../config/config.json');
+
+describe('ENSIntegration', async () => {
+    let proxyAddress: string;
+    let iexecPoco: IexecInterfaceNative;
+    let iexecPocoAsAdmin: IexecInterfaceNative;
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.ts:22]: Make it an array ?

File: test/byContract/ENSIntegration/ENSIntegration.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    async function initFixture() {
+        ({ iexecAdmin, anyone } = await getIexecAccounts());
+        const ensRegistryAddress = (await deployments.get('ENSRegistry')).address;
+        ensRegistry = ENSRegistry__factory.connect(ensRegistryAddress, anyone);
+        iexecPoco = IexecInterfaceNative__factory.connect(proxyAddress, anyone);
+        iexecPocoAsAdmin = iexecPoco.connect(iexecAdmin);
+    }
+
+    describe('Forward resolution', async () => {
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.ts:42]: I don't get the "forward" in the describe naming ?

File: test/byContract/ENSIntegration/ENSIntegration.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * Get the primary ENS name associated to the given address using reverse resolution.
+     * @param address
+     * @returns ENS name
+     */
+    async function lookup(address: string) {
+        const nameHash = ethers.utils.namehash(`${address.substring(2)}.addr.reverse`);
+        const reverseResolver = await getResolver(nameHash);
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.ts:115]: what is the difference between reverseResolver and reverse line 104

File: test/byContract/ENSIntegration/ENSIntegration.ts (hunk 4 above)
Comment (zguesmi) [reply test/byContract/ENSIntegration/ENSIntegration.ts]: "Forward" resolution is the "normal" process of going from an ENS name to an address. We generally use "resolution" but we specify "forward" or "reverse" when both are in discussion.
See:
https://docs.ens.domains/learn/resolution#forward-resolution
https://docs.ens.domains/learn/resolution#reverse-resolution

File: test/byContract/ENSIntegration/ENSIntegration.ts (hunk 5 above)
Comment (zguesmi) [reply test/byContract/ENSIntegration/ENSIntegration.ts]: Both are resovers, I specified `reverseResolver` to emphasize that it's used to do reverse resolving.

File: test/byContract/ENSIntegration/ENSIntegration.ts (hunk 4 above)
Comment (gfournieriExec) [reply test/byContract/ENSIntegration/ENSIntegration.ts]: Thanks for the resources

File: test/byContract/ENSIntegration/ENSIntegration.test.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
+// SPDX-FileCopyrightText: 2020-2024 IEXEC BLOCKCHAIN TECH <contact@iex.ec>
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.test.ts:1]: why did you change the name of the file with `test.` ?

Comment (gfournieriExec) [reply test/byContract/ENSIntegration/ENSIntegration.test.ts:1]: I thought doing the same on a previous PR but abandoned it, since we don't do it for each test file => or we should set a naming file convention

File: test/byContract/ENSIntegration/ENSIntegration.test.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    async function initFixture() {
+        ({ iexecAdmin, anyone } = await getIexecAccounts());
+        const ensRegistryAddress = (await deployments.get('ENSRegistry')).address;
+        ensRegistry = ENSRegistry__factory.connect(ensRegistryAddress, anyone);
+        iexecPoco = IexecInterfaceNative__factory.connect(proxyAddress, anyone);
+        iexecPocoAsAdmin = iexecPoco.connect(iexecAdmin);
+    }
+
+    describe('Forward resolution', () => {
+        it('Should resolve initial names', async () => {
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.test.ts:41]: Is it possible to create a should not resolve test ? like pointing to address(0) if it's the default resolve address

File: test/byContract/ENSIntegration/ENSIntegration.test.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * Get the address associated to the given ENS name using forward resolution.
+     * @param domain ENS domain name
+     * @returns ETH address
+     */
+    async function resolve(domain: string) {
+        const nameHash = ethers.utils.namehash(domain);
+        const resolver = await getResolver(nameHash);
+        return await resolver['addr(bytes32)'](nameHash);
```
Comment (gfournieriExec) [inline test/byContract/ENSIntegration/ENSIntegration.test.ts:103]: I don't get this operation

File: test/byContract/ENSIntegration/ENSIntegration.test.ts (hunk 6 above)
Comment (zguesmi) [reply test/byContract/ENSIntegration/ENSIntegration.test.ts:1]: Using `.test` prefix seems to be the convention.
I'm preparing a PR to rename other files https://github.com/iExecBlockchainComputing/PoCo/pull/106

File: test/byContract/ENSIntegration/ENSIntegration.test.ts (hunk 8 above)
Comment (zguesmi) [reply test/byContract/ENSIntegration/ENSIntegration.test.ts:103]: The steps to resolve an ENS name are:
1. Call ENS registry to get the resolver associated to the name => `getResolver()`
2. Query the resolver to get the address mapped to the name => `resolver.addr()`

I recommand you check this for in-depth explanation: https://docs.ens.domains/resolution.

The form `['addr(bytes32)']` is used to specify the exact signature of the function `addr` because there are multiple declarations (`addr(bytes32 node)`, `addr(bytes32 node, uint256 coinType)`, ...)

File: test/byContract/ENSIntegration/ENSIntegration.test.ts (hunk 6 above)
Comment (zguesmi) [reply test/byContract/ENSIntegration/ENSIntegration.test.ts:1]: https://jestjs.io/docs/configuration#testregex-string--arraystring

File: test/byContract/ENSIntegration/ENSIntegration.test.ts (hunk 8 above)
Comment (gfournieriExec) [reply test/byContract/ENSIntegration/ENSIntegration.test.ts:103]: Thanks for the informations !

**Review by gfournieriExec — APPROVED** [review]
Thanks!
